staging/src/k8s.io/component-base/logs/datapol/datapol.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+// Package datapol contains functions to determine if objects contain sensitive
+// data to e.g. make decisions on whether to log them or not.
+package datapol
+
+import (
+	"reflect"
+	"strings"
+)
+
+// Verify returns a list of the datatypes embeded in the argument that can be
+// considered sensitive w.r.t. to logging
+func Verify(value interface{}) []string {
+	t := reflect.ValueOf(value)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	return datatypes(t)
+}
+
+func datatypes(v reflect.Value) []string {
+	switch v.Kind() {
+	case reflect.Slice, reflect.Array:
+		for i := 0; i < v.Len(); i++ {
+			if types := datatypes(v.Index(i)); len(types) > 0 {
+				return types
+			}
+		}
+	case reflect.Map:
+		mapIter := v.MapRange()
+		for mapIter.Next() {
+			k := mapIter.Key()
+			v := mapIter.Value()
+			if types := datatypes(k); len(types) > 0 {
+				return types
+			}
+			if types := datatypes(v); len(types) > 0 {
+				return types
+			}
+		}
+	case reflect.Struct:
+		if types := byType(v.Type()); len(types) > 0 && !v.IsZero() {
+			return types
+		}
+		t := v.Type()
+		numField := t.NumField()
+
+		for i := 0; i < numField; i++ {
+			f := t.Field(i)
+			if f.Type.Kind() == reflect.Ptr || f.Type.Kind() == reflect.Interface {
```

Comment:
I'm not sure `Interface` kinds should be skipped.  For instance, I would include the following test case, which currently fails.

```golang
	{
		name:   "Interface value stores struct with annotated field",
		value:  struct{value interface{}}{withDatapolTag{Key: "nonempty"}},
		expect: []string{"password"},
	}
```

As discussed elsewhere, I'm also not sure on when pointers do and don't get printed as addresses rather than underlying values.  Personally, I might be inclined to always examine fields and defer the implementation of skipping pointers as a `beta` objective, if `beta` is going to contain other optimizations as well.

Reply:
Ack, not skipping interface anymore, and added some test logic to verify that we do the validation correctly

Reply:
When it comes to pointers there is an obvious problem with potential cycles. I believe that the detection of those will make implementation overly complicated and will require an extra memory. As the same problem with cycles detection applies also to formatting functions I do not think that following pointer is really needed.

Reply:
I added some test logic, and I can indeed skip pointers, but not interfaces. So this would prevent cycles. (My initial understanding was that Interfaces were always pointers, due to how almost all structs implements their functions on pointers to the struct, but even in reflection this will show up as Interface with the kind of their Elem as Pointer.